File: day1/solution.py
def part1():
    file = open("input.txt")
    left = list()
    right = list()
    for line in file.readlines():
        left.append(int(line[0:5]))
        right.append(int(line[-6:-1]))
    file.close()
    left.sort()
    right.sort()
    sum = 0
    for i in range(len(left)):
        sum += abs(left[i] - right[i])

    return sum

def part2():
    file = open("input.txt")
    left = list()
    right = dict()
    for line in file.readlines():
        left.append(int(line[0:5]))
        right[int(line[-6:-1])] = 1 + right.get(int(line[-6:-1]), 0)
    file.close()
    sum = 0
    for num in left:
        sum += num * right.get(num, 0)

    return sum

if __name__ == "__main__":
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")

File: day1/solution.go
package day1

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func abs(x int) int {
	if x < 0 {
		return x * -1
	}
	return x
}

func Part1() int {
	file, err := os.Open("day1/input.txt")
	if err != nil {
		fmt.Println("Error: ", err)
		return -1
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var left, right []int

	for scanner.Scan() {
		line := strings.Split(scanner.Text(), "   ")
		leftNum, _ := strconv.Atoi(line[0])
		rightNum, _ := strconv.Atoi(strings.TrimSpace(line[1]))
		left = append(left, leftNum)
		right = append(right, rightNum)
	}

	slices.Sort(left)
	slices.Sort(right)
	sum := 0

	for i := 0; i < len(left); i++ {
		sum += abs(left[i] - right[i])
	}
	return sum
}

type IntMap map[int]int

func (m IntMap) addHelper(num int) {
	_, exists := m[num]
	if !exists {
		m[num] = 1
	} else {
		m[num] += 1
	}
}

func (m IntMap) getHelper(num int) int {
	value, exists := m[num]
	if !exists {
		return 0
	}
	return value
}

func Part2() int {
	file, err := os.Open("day1/input.txt")
	if err != nil {
		fmt.Println("Error: ", err)
		return -1
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var left []int
	right := make(IntMap)

	for scanner.Scan() {
		line := strings.Split(scanner.Text(), "   ")
		leftNum, _ := strconv.Atoi(line[0])
		rightNum, _ := strconv.Atoi(strings.TrimSpace(line[1]))
		left = append(left, leftNum)
		right.addHelper(rightNum)
	}

	sum := 0

	for i := 0; i < len(left); i++ {
		sum += left[i] * right.getHelper(left[i])
	}

	return sum
}
